dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/internal/BNFHeadersImpl.java

```diff
@@ -122,6 +123,12 @@ public abstract class BNFHeadersImpl implements BNFHeaders, Externalizable {
 
     /** Storage for the header/value pairs */
     private transient HashMap<Integer, HeaderElement> storage = new HashMap<Integer, HeaderElement>();
+
+    /**
+     * Header keys added to this BNFHeaderImpl instance, indexed by name.
+     */
+    private final ConcurrentHashMap<String, HeaderKeys> headerKeysByName = new ConcurrentHashMap<>();
```

Comment:
Agree that headerKeysByName doesn't _need to be_ a ConcurrentHashMap, but I read somewhere that CHM is actually slightly faster than HashMap for regular operations, so I went with that instead.

Will change `headerKeysByName` to transient - wasn't sure if it really mattered, because not clear when anything in this class would be seriallized anyway - figured you would advise me. :-)

Reply:
ConcurrentHashMap is meant to scale better than a map that is synchronized via a single lock since it splits the single lock into several ones thus preventing falsely contended scenarios involving unrelated parts of the map. I doubt whether ConcurrentHashMap would be better than a normal HashMap if synchronization is not to be considered for the HashMap. Since there is extra logic/path length in ConcurrentHashMap to manage the lock splitting. I thought I'd say that ConcurrentHashMap is not guaranteed to be faster (or slower, as with most things performance related, it depends).

Reply:
What I read was that ConcurrentHashMap was slightly faster than plain HashMap even for single-threaded usage, which I found instinctively puzzling ... maybe I should not believe everything I read on the Internet?  :-)

Since stuff in a servlet is guaranteed by the container to be thread-safe (IIUC) then I can switch to HashMap and see how that performs.